### DISCUSSION TITLE
"Uncaught TypeError: $ is not a function" with Storybook & Gatsby v3.10.1

### DISCUSSION THREAD

**Discussion by vcarrara** [gatsbyjs/gatsby#32512] [Help]
### Preliminary Checks

- [x] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I'm trying to integrate Storybook with a fresh and new Gatsby project. After installing the boilerplate using `npm init gatsby`, I followed Gatsby v3 official documentation about **Visual Testing with Storybook**.

When I run `npm run storybook`, loading is **stuck at 99%** :

```
99% done plugins webpack-hot-middlewarewebpack built preview 92d6d906c27e54121f0f in 2185ms
```

I can access Storybook (localhost:6006) but the page is loading infinitely :

![image](https://user-images.githubusercontent.com/48546263/126975931-02350daa-e3f7-4164-b8bb-6fcecd84a19a.png)

In my browser's console I have this error : **"Uncaught TypeError: $ is not a function"** :

![image](https://user-images.githubusercontent.com/48546263/126975811-f5450155-7581-4aee-8678-184e983daf6b.png)

Here are my Storybook config files :

.storybook/main.js

```js
module.exports = {
  // You will want to change this to wherever your Stories will live
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: ["@storybook/addon-links", "@storybook/addon-essentials"],
  core: {
    builder: "webpack5",
  },
  webpackFinal: async config => {
    // Transpile Gatsby module because Gatsby includes un-transpiled ES6 code.
    config.module.rules[0].exclude = [/node_modules\/(?!(gatsby)\/)/]
    // use babel-plugin-remove-graphql-queries to remove static queries from components when rendering in storybook
    config.module.rules[0].use[0].options.plugins.push(
      require.resolve("babel-plugin-remove-graphql-queries")
    )
    return config
  },
}
```

.storybook/preview.js

```js
import { action } from "@storybook/addon-actions"
// Gatsby's Link overrides:
// Gatsby Link calls the `enqueue` & `hovering` methods on the global variable ___loader.
// This global object isn't set in storybook context, requiring you to override it to empty functions (no-op),
// so Gatsby Link doesn't throw errors.
global.___loader = {
  enqueue: () => {},
  hovering: () => {},
}
// This global variable prevents the "__BASE_PATH__ is not defined" error inside Storybook.
global.__BASE_PATH__ = "/"
// Navigating through a gatsby app using gatsby-link or any other gatsby component will use the `___navigate` method.
// In Storybook it makes more sense to log an action than doing an actual navigate. Checkout the actions addon docs for more info: https://github.com/storybookjs/storybook/tree/master/addons/actions
window.___navigate = pathname => {
  action("NavigateTo:")(pathname)
}
```

gatsby-config.js
```js
module.exports = {
  siteMetadata: {
    siteUrl: "https://www.yourdomain.tld",
    title: "test-gatsby-storybook",
  },
  plugins: [],
};

```

package.json
```json
{
  "name": "test-gatsby-storybook",
  "version": "1.0.0",
  "private": true,
  "description": "test-gatsby-storybook",
  "keywords": [
    "gatsby"
  ],
  "scripts": {
    "develop": "gatsby develop",
    "start": "gatsby develop",
    "build": "gatsby build",
    "serve": "gatsby serve",
    "clean": "gatsby clean",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "dependencies": {
    "gatsby": "^3.10.1",
    "react": "^17.0.1",
    "react-dom": "^17.0.1"
  },
  "devDependencies": {
    "@babel/core": "^7.14.8",
    "@storybook/addon-actions": "^6.3.6",
    "@storybook/addon-essentials": "^6.3.6",
    "@storybook/addon-links": "^6.3.6",
    "@storybook/builder-webpack5": "^6.3.6",
    "@storybook/manager-webpack5": "^6.3.6",
    "@storybook/react": "^6.3.6",
    "babel-loader": "^8.2.2"
  }
}
```

### Reproduction Link

https://github.com/vcarrara/test-gatsby-storybook

### Steps to Reproduce

1. Run `npm run storybook`

### Expected Result

Storybook UI should start correctly and I should be able to see stories defined in `src` folder

### Actual Result

- Storybook cli is stuck at 99%
- Storybook UI is loading infinitely with a "**Uncaught TypeError: $ is not a function**" error in the browser's console

### Environment

```shell
System:
  OS: Windows 10 10.0.19042
  CPU: (8) x64 Intel(R) Core(TM) i5-8300H CPU @ 2.30GHz
Binaries:
  Node: 12.13.0 - C:\Program Files\nodejs\node.EXE
  npm: 6.12.0 - C:\Program Files\nodejs\npm.CMD
Languages:
  Python: 3.7.9
Browsers:
  Chrome: 92.0.4515.107
  Edge: Spartan (44.19041.1023.0), Chromium (92.0.902.55)
npmPackages:
  gatsby: ^3.10.1 => 3.10.1
```

**Comment by doctorpangloss** [comment]
I have the same issue.

**Comment by fbuireu** [comment]
I would suggest adding `core-js` as excluded module and resolving the following fields `browser`, `module`, `main`. So your final configuration may look like something:

```javascript
module.exports = {
  // You will want to change this to wherever your Stories will live
  stories: ["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"],
  addons: ["@storybook/addon-links", "@storybook/addon-essentials"],
  core: {
    builder: "webpack5",
  },
  webpackFinal: async config => {
    // Transpile Gatsby module because Gatsby includes un-transpiled ES6 code.
    config.module.rules[0].exclude = [/node_modules\/(?!(gatsby)\/)/]
    config.module.rules[0].exclude = [/core-js/] // new line

    // use babel-plugin-remove-graphql-queries to remove static queries from components when rendering in storybook
    config.module.rules[0].use[0].options.plugins.push(
      require.resolve("babel-plugin-remove-graphql-queries")
    )

    config.resolve.mainFields = ["browser", "module", "main"] // new line
    return config
  },
}
```

Of course, you can concatenate both excluding (`exclude`) lines.

**Reply by unimprobable** [reply]
@fbuireu This solution worked to get past the initial issue of the "Uncaught TypeError: $ is not a function", but it stripped my Tailwind implementation. Any idea why or how to get that back? I am using
`import '../src/styles/global.css'`
in the .storybook/preview.js file.
Thanks!

**Reply by unimprobable** [reply]
@fbuireu Nevermind, I did a little more digging and realized I needed to add @storybook/addon-postcss addon to get Tailwind to work with Storybook and Gatsby.

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-implicit-postcss-loader

Thanks!
